File: go/protein-translation/protein_translation.go
package protein

const testVersion = 1

var codonMap = map[string]string{
	"AUG": "Methionine",
	"UUU": "Phenylalanine",
	"UUC": "Phenylalanine",
	"UUA": "Leucine",
	"UUG": "Leucine",
	"UCU": "Serine",
	"UCC": "Serine",
	"UCA": "Serine",
	"UCG": "Serine",
	"UAU": "Tyrosine",
	"UAC": "Tyrosine",
	"UGU": "Cysteine",
	"UGC": "Cysteine",
	"UGG": "Tryptophan",
	"UAA": "STOP",
	"UAG": "STOP",
	"UGA": "STOP",
}

func FromRNA(input string) (proteins []string) {

	var pos int
	for {

		if pos+3 > len(input) {
			return
		}
		codon := FromCodon(input[pos : pos+3])
		if codon == "STOP" {
			return
		}
		proteins = append(proteins, codon)
		pos += 3

	}
}

func FromCodon(input string) string {
	return codonMap[input]
}

File: haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation
  ( proteins
  ) where

import           Data.List.Split

proteins :: String -> Maybe [String]
proteins s
  | length s `rem` 3 /= 0 = Nothing
  | otherwise = Just $ takeWhile (/= "STOP") $ map protein $ chunksOf 3 s

protein :: String -> String
protein s
  | s `elem` ["AUG"] = "Methionine"
  | s `elem` ["UUU", "UUC"] = "Phenylalanine"
  | s `elem` ["UUA", "UUG"] = "Leucine"
  | s `elem` ["UCU", "UCC", "UCA", "UCG"] = "Serine"
  | s `elem` ["UAU", "UAC"] = "Tyrosine"
  | s `elem` ["UGU", "UGC"] = "Cysteine"
  | s `elem` ["UGG"] = "Tryptophan"
  | s `elem` ["UAA", "UAG", "UGA"] = "STOP"
  | otherwise = ""
